freqtrade/edge/__init__.py

```diff
@@ -0,0 +1,388 @@
+# pragma pylint: disable=W0603
+""" Edge positioning package """
+import logging
+from typing import Any, Dict
+import arrow
+
+import numpy as np
+import utils_find_1st as utf1st
+from pandas import DataFrame
+
+import freqtrade.optimize as optimize
+from freqtrade.optimize.backtesting import BacktestResult
+from freqtrade.arguments import Arguments
+from freqtrade.exchange import Exchange
+from freqtrade.strategy.interface import SellType
+from freqtrade.strategy.resolver import IStrategy, StrategyResolver
+from freqtrade.optimize.backtesting import Backtesting
+
+
+logger = logging.getLogger(__name__)
+
+
+class Edge():
+
+    config: Dict = {}
+
+    def __init__(self, config: Dict[str, Any], exchange=None) -> None:
+        """
+        constructor
+        """
+        self.config = config
+        self.strategy: IStrategy = StrategyResolver(self.config).strategy
+        self.ticker_interval = self.strategy.ticker_interval
+        self.tickerdata_to_dataframe = self.strategy.tickerdata_to_dataframe
+        self.get_timeframe = Backtesting.get_timeframe
+        self.advise_sell = self.strategy.advise_sell
+        self.advise_buy = self.strategy.advise_buy
+
+        self.edge_config = self.config.get('edge', {})
+
+        self._last_updated = None
+        self._cached_pairs: list = []
+        self._total_capital = self.edge_config['total_capital_in_stake_currency']
+        self._allowed_risk = self.edge_config['allowed_risk']
+
+        ###
+        #
+        ###
+        if exchange is None:
+            self.config['exchange']['secret'] = ''
+            self.config['exchange']['password'] = ''
+            self.config['exchange']['uid'] = ''
+            self.config['dry_run'] = True
+            self.exchange = Exchange(self.config)
+        else:
+            self.exchange = exchange
+
+        self.fee = self.exchange.get_fee()
+
+    def calculate(self) -> bool:
+        pairs = self.config['exchange']['pair_whitelist']
+        heartbeat = self.config['edge']['process_throttle_secs']
+
+        if (self._last_updated is not None) and (
+                self._last_updated + heartbeat > arrow.utcnow().timestamp):
+            return False
+
+        data: Dict[str, Any] = {}
+        logger.info('Using stake_currency: %s ...', self.config['stake_currency'])
+        logger.info('Using stake_amount: %s ...', self.config['stake_amount'])
+        logger.info('Using local backtesting data (using whitelist in given config) ...')
+        # TODO: add "timerange" to Edge config
+        timerange = Arguments.parse_timerange(None if self.config.get(
```

Comment:
i suspect parsing the timerange in the constructor will be enough - doesn't seem to make sense to do that each iteration ...